text/0007-performance-data-propagation.md

```diff
@@ -0,0 +1,179 @@
+
+# Motivation
+[motivation]: #motivation
+
+This is needed to provide both a default and user specific implementations for future Network Profilers and Network Stats overlays. The MLAPI source code can remain Open Source, and yet, the data being streamed can be hooked up to external and low level systems.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In order to allow first and third party rendering of the data we will need to allow both MLAPI and its transports to propagate data up the chain. 
+
+If the profiler flags are enabled the following will need to happen:
+
+Transports will need to output the number of bytes being sent in the current tick and the latency. The data will be written out to a `ProfilerTickData` chunk. This data will be generic and transport agnostic.
+
+MLAPI will need to output the number of networked objects and the amount of RPCs being sent. This data will be added to the `ProfilerTickData` chunk.
+
+External Tools and APIs will be able to then get this fully written `ProfilerTickData` at the end of the frame for them to do what they wish (Render to Profiler Stats Overlay or Display/Record in a Profiler or file)
+
+As this data is largely a `Dictionary`, accessing its data will be as simple as accessing things by specified fields. Additionally the rendering or logging of these statistics will fall on the user and not affect the Tick or be done at all within the MLAPI framework.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+In order to limit the interface, we are pushing the collected set or performance data to be obtainable only at the end of a tick. The following is roughly what this object could look like. Currently this object will be updated in `NetworkManager.LateUpdate()` for the MLAPI and `Transport.Send()` and other relevant transport method for the transport layer. At the end of the tick the user will be notified that the relevant data is complete and ready (both the MLAPI data and Transport data). Since MLAPI is responsible for collecting the data, and the data from transports will be generic and apply to all/most transports, MLAPI as  the owner of the collection of performance data will call `GetTransportGetData()` on the transports through an interface in order to get the data exactly when it needs it in the pipeline before propagating the data out externally. Whether or not the data is collected per transport can also be toggled by a runtime flag `profilerEnabled`.
+
+## Pseudo-code
+
+```cs
+namespace MLAPI
+{
+    class ProfilerConstants
+    {
+        public const string NumberOfServerRPCs = "numberOfServerRPCs";
```

Comment:
I recently went through an exercise to clean up the channel stuff that might be of use here.  Check out Transport.cs circa line 9 / the 'Channel' definition.  You can define each measurement as an enum, then if you need to output a string you can just call ToString() on it.  Then you don't have to maintain the strings / pay for the string de-ref.

Reply:
From another comment by @becksebenius-unity :

"Suggestion for the dictionary to use string keys instead of a predefined enum. unintuitively, it's actually better for performance (enums as dictionary keys actually get boxed 🤢). and if we're going to go with a flexible approach, lets maximize that instead of restricting them to predefined keys, that way users building their own tools for their own custom transport could extend. we can still provide predefined keys for the user in a constants class so that we can guarantee good data for our profiler."

Reply:
I think he's not suggesting to change `Dictionary<string, int>` but to define an enum for "our" `ProfilerConstants` to be an `enum` and do `ProfilerConstants. NumberOfServerRPCs.ToString()` instead of string constants.

Reply:
Ah i see. I can update to that. in the actual PR